Add Usage Plan to Amazon API Gateway

laardee opened the pull request:
<!--
1. Please check out and follow our Contributing Guidelines: https://github.com/serverless/serverless/blob/master/CONTRIBUTING.md
2. Do not remove any section of the template. If something is not applicable leave it empty but leave it in the PR
3. Please follow the template, otherwise we'll have to ask you to update it and it will take longer until your PR is merged
-->

## What did you implement:

This PR adds usage plan to Amazon API Gateway.

Closes #2450
Closes #2741

<!--
Briefly describe the feature if no issue exists for this PR
-->

## How did you implement it:
By adding `AWS::ApiGateway::UsagePlan` and `AWS::ApiGateway::UsagePlanKey` to Amazon API Gateway build flow.
<!--
If this is a nontrivial change please briefly describe your implementation so its easy for us to understand and review your code.
-->

## How can we verify it:

Create a service with API key(s) and verify that deployment creates usage plan and usage plan key(s) to CF stack and AWS.

<!--
Add any applicable config, commands, screenshots or other resources
to make it easy for us to verify this works. The easier you make it for us
to review a PR, the faster we can review and merge it.

Examples:
* serverless.yml - Fully functioning to easily deploy changes
* Screenshots - Showing the difference between your output and the master
* AWS CLI commands - To list AWS resources and show that the correct config is in place
* Other - Anything else that comes to mind to help us evaluate
-->

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

laardee commented on lib/plugins/aws/remove/lib/stack.js:
```diff
@@ -26,8 +27,47 @@ module.exports = {
       .then(() => cfData);
   },
 
+  disassociateUsagePlan() {
```
@pmuens, about this. It seems that it is not possible to remove usage plan that that is associated with deployment stage using CloudFormation delete stack method. The usage plan has to be made after deployment and stage to get a reference to the stage. So delete stack tries to remove the usage plan before the stage and that cannot be done if it is associated with the stage 🤔.

laardee replied:
If this is an OK approach to remove the association from usage plan, I'll change first two requests (CloudFormation describeStackResource and APIGateway getUsagePlans) to be parallel so it should be a little bit faster.

pmuens replied:
🤔 Haven't had time to take a deep dive into it yet. Will do this next week!

So it's not possible to completely rely on CloudFormation here? Even if `DependsOn` are added to the resources to declare their dependencies explicitly?

laardee replied:
I can give it another try later, but I'm afraid that the order of the removal is still wrong.

pmuens replied:
Sounds good. Thanks for looking into it. 👍 

I'll also take a deeper dive into it next week. Maybe we can uncover something there...

laardee replied:
@pmuens I created an example CF stack and added [a question](https://forums.aws.amazon.com/thread.jspa?messageID=787503&#787503) about removal to the AWS Developer Forum.

It seems that it is not normal behavior that Usage Plan prevents stack deletion and they will look into it. I don't know how soon they can provide a solution, though.

pmuens replied:
> @pmuens I created an example CF stack and added a question about removal to the AWS Developer Forum. It seems that it is not normal behavior that Usage Plan prevents stack deletion and they will look into it. I don't know how soon they can provide a solution, though.

Great @laardee 👍 !

Thanks for researching! Glad to hear that it's an unusual behavior since this could make the implementation way easier (once the problem is resolved).

laardee commented:
@pmuens, about the usage plan quota and throttle, should I add it to provider 
```yaml
provider:
  name: aws
  runtime: nodejs6.10
  apiKeys:
    - api-key-first
    - api-key-second
  usagePlan:
    quota:
      limit: Integer
      offset: Integer
      period: String
    throttle:
      burstLimit: Integer
      rateLimit: Number
```
or is it something that could be only overwritable in resources
```yaml
resources:
  Resources:
    ApiGatewayUsagePlan:
      Type: AWS::ApiGateway::UsagePlan
      DependsOn: ApiGatewayStage
      Properties:
        Quota:
          Limit: Integer
          Offset: Integer
          Period: String
        Throttle:
          BurstLimit: Integer
          RateLimit: Number
```

pmuens commented:
@laardee Thanks for working on and updating this! Really excited about this feature since it is unfortunately broken for a long time due to lack of CloudFormation support.

> @pmuens, about the usage plan quota and throttle, should I add it to provider

I would vote to add it to the `provider` section since the `apiKeys` are already there 👍

pmuens commented on lib/plugins/aws/package/compile/events/apiGateway/lib/deployment.js:
```diff
@@ -13,7 +13,6 @@ module.exports = {
         Type: 'AWS::ApiGateway::Deployment',
         Properties: {
           RestApiId: { Ref: this.apiGatewayRestApiLogicalId },
-          StageName: this.options.stage,
```
Isn't this removal introducing a breaking change for developers relying on this information?

laardee replied:
The stage is now created with Stage resource, it has the same stage name. I'm not sure if it makes any difference if the stage is created with the resource or automatically under the hood.

HyperBrain replied:
I also create separate stage objects in my alias plugin to be able to have multiple alias deployments within one stage.
From my experience I can say, that the only thing that will happen with this change is, that you probably have to manually delete the stage that was auto-generated with the previous deployment when switching to separate stage resources. (CF will tell you that the stage already exists). But this would only be a one time action.
The other thing is, that the stage must be made dependent on the deployment.
FYI: Having a dedicated stage object is also the only proper way to offer cache configuration and APIG logging configuration for deployed APIs and would be needed anyway in the future.

See my code here: https://github.com/HyperBrain/serverless-aws-alias/blob/master/lib/stackops/apiGateway.js#L77-L94

pmuens commented on lib/plugins/aws/lib/naming.js:
```diff
@@ -182,6 +182,15 @@ module.exports = {
   getApiKeyLogicalIdRegex() {
     return /^ApiGatewayApiKey/;
   },
+  getUsagePlanLogicalId() {
```
Would be great to have unit tests for those functions.

laardee replied:
Done ✅

pmuens commented on lib/plugins/aws/lib/naming.js:
```diff
@@ -182,6 +182,15 @@ module.exports = {
+  getUsagePlanLogicalId() {
+    return 'ApiGatewayUsagePlan';
+  },
+  getUsagePlanKeyLogicalId(usagePlanKeyNumber) {
```
The rules for the generation of those logical ids and some examples should be added to the docs (https://github.com/serverless/serverless/blob/master/docs/providers/aws/guide/resources.md#aws-cloudformation-resource-reference).

laardee replied:
Done ✅ I also added one sentence about usage plan quota and throttle to the [API Gateway documentation](https://github.com/serverless/serverless/pull/3715/files#diff-871929e390d61a5d2a3b505ac9b4bf24).

pmuens commented:
Hey @laardee just wanted to do a quick follow-up here.

Have you heard anything from AWS yet? Do you need any assistant here?

laardee commented:
@pmuens, no, nothing yet from AWS 😞. I'll ask if they have made any progress.

pmuens commented:
> no, nothing yet from AWS 😞. I'll ask if they have made any progress.

Alright. Thanks for the update @laardee 

I'll keep an eye on this PR so that we can get it into a GTM state ASAP.

laardee commented:
@pmuens I added the usage plan to the provider object and fixed the small issues. I think all the features for this PR are now about ready. Now we just have to wait what AWS says about the removal issue.

pmuens commented:
> @pmuens I added the usage plan to the provider object and fixed the small issues. I think all the features for this PR are now about ready. Now we just have to wait what AWS says about the removal issue.

Great! Thanks for the update @laardee 💪 

Yes, would be nice if we could get that into this PR. I'll take another deep dive into this again starting next week 👍

laardee commented:
@pmuens still no news from AWS. Should we continue with some kind of plan B or what should we do?

pmuens commented:
> @pmuens still no news from AWS. Should we continue with some kind of plan B or what should we do?

Thanks for the follow-up @laardee 👍 

🤔 yes, we should have a Plan B here. Maybe we could proceed with the current solution if AWS won't provide some feedback until Friday.

I'll review this PR once again at the end of the week to make sure that everything is alright.

Hopefully we can merge it with the fixes for AWS (otherwise we should just merge the current implementation).

laardee commented:
@pmuens nice, thanks. Hopefully the CloudFormation support will be available soon, I'll create a new issue when that happens.

pmuens commented:
> @pmuens nice, thanks. Hopefully the CloudFormation support will be available soon, I'll create a new issue when that happens.

Thanks for the PR @laardee 🎉 Yes, we should refactor that later on, but it's working fine with this workaround for now 💯

eahefnawy commented:
Hey @laardee ... thanks a lot for this feature 🙌 

I've tested it with an existing service and it's throwing an "stage already exist" error just like @HyperBrain mentioned, which would break pretty much all our users. So I dropped the new stage resource and kept creating the stage at the deployment resource and things are now working again.

Any reason why you felt the need to create a stage resource rather than created at deployment as it is?

laardee commented:
@eahefnawy The main reason was why I made the change was that I thought that it I would get more control of the removing it when I have the stage defined as a CloudFormation resource.

HyperBrain commented:
I suggest to follow @eahefnawy 's proposal to just ref the deployment here. The damage/benefit ratio is just too big with introducing separate stage resources in regards to existing deployments.

For separate stages - users can use the alias plugin if they want more control over the stages. I'm currently implementing detailed stage settings there, where it is much easier than within the core Serverless framework. BTW: I'll add APIKey support with usage plans in the near future, so that your additions will work fine there and even will allow different usage plans per alias/APIG stage deployed simultaneously 😄

pmuens commented:
> @eahefnawy The main reason was why I made the change was that I thought that it I would get more control of the removing it when I have the stage defined as a CloudFormation resource.

Thanks for the insights @laardee 👍 

We'll follow drop the stage resource for now (as @eahefnawy mentioned) to introduce it in a non-breaking way. Furthermore thanks @HyperBrain for providing a potential solution to use stage resources with the help of the alias plugin.

pmuens commented:
Closing this since @eahefnawy PRed a follow up PR with #3819 

Feel free to comment if necessary. Thanks again @laardee for your hard work on this feature 🙌 💯

antanasbrazenas commented:
I'm not sure current disassociation works when removing a stack. CloudFormation results in "`Cannot delete Usage Plan ****** because there are API Stages associated with it, including...`" error message despite of serverless console starting removal with "`Removing usage plan association...`" log.

pmuens commented:
Hey @antanasbrazenas thanks for commenting 👍

Could you open up a new issue for that (with steps on how to reproduce this)? Thanks in advance!
